skrifa/src/scale/cff/hint.rs

```diff
@@ -904,6 +907,182 @@ fn msb_mask(bit: usize) -> u8 {
+        self.stem_count = index as u8 + 1;
+    }
+
+    fn build_hint_map(&mut self, mask: Option<HintMask>, origin: Fixed) {
+        self.map.build(
+            self.state,
+            mask,
+            Some(&mut self.initial_map),
+            &mut self.stem_hints[..self.stem_count as usize],
+            origin,
+            false,
+        );
+    }
+}
+
+impl<'a, S: CommandSink> CommandSink for HintingSink<'a, S> {
+    fn hstem(&mut self, min: Fixed, max: Fixed) {
+        self.add_stem(min, max);
+    }
+
+    fn hint_mask(&mut self, mask: &[u8]) {
+        // For invalid hint masks, FreeType assumes all hints are active.
+        // See <https://gitlab.freedesktop.org/freetype/freetype/-/blob/80a507a6b8e3d2906ad2c8ba69329bd2fb2a85ef/src/psaux/pshints.c#L844>
+        let mask = HintMask::new(mask).unwrap_or_else(|| HintMask::all());
+        if mask != self.mask {
+            self.mask = mask;
+            self.map.is_valid = false;
+        }
+    }
+
+    fn counter_mask(&mut self, mask: &[u8]) {
+        // For counter masks, we build a temporary hint map "just to
+        // place and lock those stems participating in the counter
+        // mask." Building the map modifies the stem hint array as a
+        // side effect.
+        // See <https://gitlab.freedesktop.org/freetype/freetype/-/blob/80a507a6b8e3d2906ad2c8ba69329bd2fb2a85ef/src/psaux/psintrp.c#L2617>
+        let mask = HintMask::new(mask).unwrap_or_else(|| HintMask::all());
+        let mut map = HintMap::new(self.state.scale);
+        map.build(
+            self.state,
+            Some(mask),
+            Some(&mut self.initial_map),
+            &mut self.stem_hints[..self.stem_count as usize],
+            Fixed::ZERO,
+            false,
+        );
+    }
+
+    fn move_to(&mut self, x: Fixed, y: Fixed) {
+        self.maybe_close_subpath();
+        self.start_point = Some([x, y]);
+        let x = self.scale(x);
+        let y = self.hint(y);
+        self.sink.move_to(x, y);
+    }
+
+    fn line_to(&mut self, x: Fixed, y: Fixed) {
+        self.flush_pending_line();
+        let ds_x = self.scale(x);
+        let ds_y = self.hint(y);
+        self.pending_line = Some([x, y, ds_x, ds_y]);
+    }
+
+    fn curve_to(&mut self, cx1: Fixed, cy1: Fixed, cx2: Fixed, cy2: Fixed, x: Fixed, y: Fixed) {
+        self.flush_pending_line();
+        let cx1 = self.scale(cx1);
+        let cy1 = self.hint(cy1);
+        let cx2 = self.scale(cx2);
+        let cy2 = self.hint(cy2);
+        let x = self.scale(x);
+        let y = self.hint(y);
+        self.sink.curve_to(cx1, cy1, cx2, cy2, x, y);
+    }
+
+    fn close(&mut self) {
+        // We emit close commands based on the sequence of moves.
+        // See `maybe_close_subpath`
+    }
+}
+
+/// FreeType returns values in 26.6 but we keep them in 16.6. This
+/// simply zeroes out the low 10 bits so that we match.
+fn trunc(value: Fixed) -> Fixed {
+    Fixed::from_bits(value.to_bits() & !0x3FF)
+}
```

Comment:
So to clarify, we're using `Fixed` to represent `26.6`? I vaguely recall this, but has there been any thought towards a `F26Dot6` type?

Reply:
The actual value here is 16.16 but we zero the low 10 bits so that conversion to float has the same precision as if it was 26.6. 

We do have `F26Dot6` now but we can’t actually implement it correctly for our purposes (see #232). The unfortunate reality is that FT works (sometimes) in 26.6 but uses 16.16 mul/div functions on carefully crafted values to get the expected results. If we try to do the principled thing, we’ll break compatibility. In that case, I’d rather just move to floating point (someday!)

Reply:
^ seems worth capturing in comments. Also we should file a move to float issue if we haven't already.

Reply:
There’s already a comment on this function that was cut off in the review. I’ll add the note about conversion to float to clarify the purpose. 

+1 on the issue. Will do.